src/main/java/com/vi/counselingtoolsservice/api/controller/ProxyController.java

```diff
@@ -0,0 +1,133 @@
+    }
+
+    String role = budibaseProxyService
+        .extractRolesOfCurrentUsers(request.getQueryString(), request.getHeader("cookie"));
+
+    if (role.equals("admin")) {
+      return executeNonModifiedRequest(body, method, request);
+    } else if (role.equals("consultant")) {
+      return executeConsultantRequest(body, method, request);
+    } else if (role.equals("user")) {
+      return executeUserRequest(body, method, request);
+    }
+
+    log.error(
+        "User request is not whitelisted, neither the logged in user is of role admin, consultant, user");
+    throw new IllegalStateException();
+  }
+
+  private ResponseEntity executeNonModifiedRequest(String body, HttpMethod method,
+      HttpServletRequest request) {
+    HttpHeaders headers = new HttpHeaders();
+    Enumeration<String> headerNames = request.getHeaderNames();
+    while (headerNames.hasMoreElements()) {
+      String headerName = headerNames.nextElement();
+      if (headerName.equals("accept-encoding")) {
+        headers.set("accept-encoding", "identity");
+        continue;
+      }
+      headers.set(headerName, request.getHeader(headerName));
+    }
+    return execute(request, method, body, headers);
+  }
+
+  private ResponseEntity executeConsultantRequest(String body, HttpMethod method,
+      HttpServletRequest request) {
+    budibaseProxyService.validateConsultantRequest(body, method, request);
+    HttpHeaders headers = prepareHeadersForNonAdminUser(request);
+    return execute(request, method, body, headers);
+  }
+
+  private ResponseEntity executeUserRequest(String body, HttpMethod method,
+      HttpServletRequest request) {
+    budibaseProxyService.validateUserRequest(body, method, request);
+    HttpHeaders headers = prepareHeadersForNonAdminUser(request);
+    return execute(request, method, body, headers);
+  }
+
+  private HttpHeaders prepareHeadersForNonAdminUser(HttpServletRequest request) {
+    HttpHeaders headers = new HttpHeaders();
+    Enumeration<String> headerNames = request.getHeaderNames();
+    while (headerNames.hasMoreElements()) {
+      String headerName = headerNames.nextElement();
+      if (headerName.equals("accept-encoding") || headerName.equals("cookie")) {
+        headers.set("accept-encoding", "identity");
+        continue;
+      }
+      headers.set(headerName, request.getHeader(headerName));
+    }
+
+    headers.set("x-budibase-api-key", budibaseApiKey);
+    return headers;
+  }
+
+  private ResponseEntity execute(HttpServletRequest request, HttpMethod method, String body,
+      HttpHeaders headers) {
+    String requestUrl = request.getRequestURI();
+    URI uri = null;
+    try {
+      uri = new URI("http", null, proxyServiceHost, proxyServicePort, null, null, null);
+      uri = UriComponentsBuilder.fromUri(uri)
+          .path(requestUrl)
+          .query(request.getQueryString())
+          .build(true).toUri();
+    } catch (URISyntaxException e) {
+      //
```

Comment:
missing proper error handling. At least it should be logged

Reply:
i will throw new BadRequestException();